secnet.8
.\" Man page for secnet.
.\"
.\" See the secnet.git README, or the Debian copyright file, for full
.\" list of copyright holders.
.\"
.\" secnet is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\" 
.\" secnet is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\" 
.\" You should have received a copy of the GNU General Public License
.\" version 3 along with secnet; if not, see
.\" https://www.gnu.org/licenses/gpl.html.
.TH secnet 8

.SH NAME
secnet \- VPN router daemon

.SH SYNOPSIS
\fBsecnet\fR [\fIOPTIONS\fR]

.SH DESCRIPTION
\fBsecnet\fR allows virtual private networks to be constructed
spanning multiple separate sites.

.SH OPTIONS
.TP
.B --verbose\fR, \fB-v
Enable extra diagnostics.
.TP
.B --nowarnings\fR, \fB-w
Suppress warnings.
.TP
.B --help
Display usage message.
.TP
.B --version
Display version string.
.TP
.B --nodetach\fR, \fB-n
Don't go into background.
The default behaviour is to become a daemon during startup.
.TP
.B --silent\fR, \fB--quiet\fR, \fB-f
Suppress error messages.
.TP
.B --debug\fR, \fB-d
Enable debug messages.
.TP
.B --config\fR, \fB-c \fIPATH
Specify configuration file.
The default is \fI/etc/secnet/secnet.conf\fR.
.TP
.B --just-check-config\fR, \fB-j
Check configuration and exit.
.TP
.B --sites-key\fR, \fB-s \fIKEY
Configuration file key defining active sites.
The default is \fBsites\fR.

.SH "CAPABILITY NEGOTIATION"
Sites negotiate with each other during key exchange
in order to determine which cryptographic algorithms and other features
\(en termed
.I capabilities
\(en
they each support.
Capabilities are assigned small integer numbers.
In many cases,
capability numbers can be assigned in the configuration file,
as described below;
but secnet's default assignments will often be satisfactory.
.PP
Capability numbers between 0 and 7 inclusive
are reserved for local use:
secnet will never make use of them without explicit configuration.
This may be useful to migrate from one set of parameters
for a particular cryptographic algorithm
to different, incompatible, parameters for the same algorithm.
Other capability numbers are assigned by default
by various kinds of closures.
See the descriptions below for details.
.PP
It is essential that a capability number mean the same thing
to each of a pair of peers.
It's possible to configure a site
so that it uses different capability numbers for the same feature
when it communicates with different peer sites,
but this is likely to be more confusing than useful.

.SH "CONFIGURATION FILE"
.SS Overview
The default configuration file is \fI/etc/secnet/secnet.conf\fR.
This can be overridden with the \fB--config\fR option.
.PP
The configuration file defines a dictionary (a mapping from keys to
values) of configuration information for secnet.
It is recursive in nature, i.e. values may themselves include dictionaries.
Any node in the nested structure thus defined can be identified by a
\fIpath\fR, which is the sequence of keys necessary to reach it from
the root, separated by "/" characters.
See \fBPaths\fR below for how this is used.
.PP
Furthermore, when a key is looked up in a dictionary, if it cannot be
found, it is sought in the parent dictionary, and so on back to the
root.
For instance, each \fIsite\fR must contain the \fBresolver\fR key, but
in a typical configuration there is no value in having different
resolvers for each site.
Therefore \fBresolver\fR is defined at the root and thus automatically
incorporated into all sites.
.SS Whitespace
Whitespace, including newlines, is ignored except to the extent that
it bounds other symbols.
.PP
Comment begin with "#" and continues to the end of the line.
Comments are ignored.
.SS Inclusion
A file may be recursively included into the configuration file using a
line of the form:
.IP
\fBinclude \fIPATH
.PP
This is handled at a higher level than the main parser and so
precludes the possibility of using the string \fBinclude\fR for any
other purpose.
.\" check if this is true.  it's probably a bug!
.SS Assignments
The configuration file contains one or more assigments.
Each assignment is written:
.IP
\fIkey\fR [\fB=\fR] \fIlist\fR\fB;\fR
.PP
i.e. the equals sign is optional.
The semicolon is mandatory in all contexts.
.PP
Keys start with a letter or "_" and continue with any numbers of
letters, digits, "_" and "-".
.PP
Each \fIkey\fR is a list of one or more \fIvalues\fR, separated by commas.
Possible values types are \fIboolean\fR, \fIstring\fR, \fInumber\fR,
\fIdictionary\fR, \fIpath\fR and \fIclosure evaluation\fR.
.\" This man page draws a distinction between a closure (the thing
.\" evaluated) and a closure evaluation (the closure plus is
.\" arguments).
.SS "Strings"
Strings are contained within "double quotes".
There is (currently) no escape syntax and no way to include quotes
inside strings.
.PP
Example:
.nf
        filename "/var/log/secnet";
.fi
.SS "Numbers"
Numbers are encoded in decimal and do not include a sign.
Numbers must lie in the range 0 to 4294967295.
.PP
Example:
.nf
        mtu 1400;
.fi
.SS "Dictionaries"
.\" In conffile.y dictionaries can be preceded by a search path, but
.\" this is not implemented elsewhere, so not documented here.
Dictionaries consist of one or more assignments, in the same syntax as
given above, enclosed in "{" and "}".
.PP
Example:
.nf
        system {
                userid "secnet";
                pidfile "/var/run/secnet.pid";
        };
.fi
.SS "Paths"
Paths allow a key already defined in the configuration to be aliased.
.PP
Paths consist of a sequence of keys separated by "/".
If the path starts with a "/" then it is an \fIabsolute path\fR and
the search starts at the root of the configuration.
Otherwise it is a \fIrelative path\fR and starts in the containing
dictionary or in any of its parents, down to and including the root.
If there is more than one match, the one furthest from the root "wins".
.PP
The value of a path is the list assigned to the key it refers to.
Lists are flattened; for example if a key is defined as a list of two
paths, and each of those refers to a list of two integers, the
original key is therefore defined to be a list of four integers, not
a list consisting of two lists.
.PP
It is not possible to refer to a \fIlater\fR key using a path.
.PP
Example:
.nf
        vpn {
          test {
            kakajou vpn-data/test/kakajou/kakajou;
            araminta vpn-data/test/araminta/araminta;
            deodand vpn-data/test/deodand/deodand;
            all-sites kakajou,araminta,deodand;
          };
        };
        all-sites vpn/test/all-sites;
.fi
.PP
Here, each of \fBvpn/test/kakajou\fR, \fBvpn/test/araminta\fR and
\fBvpn/test/deodand\fR are defined as aliases to values defined
elsewhere.
\fBvpn/tests/all-sites\fR is defined as the list of all three of those
values, and \fBall-sites\fR is then defined to be an alias for that.
.SS "Booleans"
The (single-element) paths \fBfalse\fR, \fBno\fR and \fBnowise\fR are
predefined and refer to a boolean false value.
Similarly \fBtrue\fR, \fByes\fR and \fBverily\fR point at a boolean
true value.
.PP
In all six cases, variants with just the first letter capitalized, and
with all letters capitalized, are also provided.
.PP
Example:
.nf
        random randomfile("/dev/urandom",no);
.fi
.SS "Closure Evaluation"
Closure evaluation uses the following syntax:
.IP
\fICLOSURE \fB( \fIARGUMENTS \fB)
.PP
\fICLOSURE\fR may be a path referring to a closure, or may itself be a
closure evaluation.
.PP
\fIARGUMENTS\fR is a list of zero or more values, separated by commas.
As a shortcut, if the arguments consist of a single dictionary, the
parentheses may be ommitted:
.IP
\fICLOSURE \fB{ \fR... \fB}
.PP
Example:
.nf
        sites map(site, vpn/test/all-sites);
.fi
.PP
When a closure is evaluated it returns a value (a list, much as above)
and may also have side effects (which may be immediate or may be
deferred to some later phase of execution).
A list of built-in closures is given below.
.SS "Mandatory Keys"
Two keys are mandatory.
\fBsystem\fR must be a dictionary in which the following keys can be
looked up:
.TP
.B log
A \fIlog closure\fR; see the \fBlogfile\fR documentation below.
The destination for log messages.
Mandatory.
.TP
.B userid
A string.
The userid to run as after dropping privilege.
Optional.
.TP
.B pidfile
A string.
The path to write a pidfile.
Optional.
.PP
\fBsites\fR should be a list of \fIsite closures\fR; see the \fBsite\fR documentation below.
This defines the collection of tunnel endpoints that \fBsecnet\fR will
communicate with.
.PP
Recall the recursive lookup logic described in \fBOverview\fR above:
if (for instance) \fBlog\fR is defined in the top level dictionary but
not in \fBsystem\fR, it will nevertheless be found when looked up in
the latter.

.SH CLOSURES
\fBsecnet\fR contains a collection of built-in closures
with names (i.e. single-element paths) given below.
.PP
Most of them return anonymous closures of various types,
which are described contextually.

.SS adns
\fBadns(\fIDICT\fB)\fR => \fIresolver closure\fR
.TP
.I DICT
This either be empty or contain the single key \fBconfig\fR, with a
string value giving configuration to supply to ADNS.
This might be read from a file using \fBreadfile\fR.
.PP
A \fIresolver closure\fR is a means of converting hostnames into
network addresses.

.SS diffie-hellman
.PP
\fBdiffie-hellman(\fIMODULUS\fB, \fIGENERATOR\fR[\fB, \fICHECK\fR]\fB)\fR => \fIdh closure\fR
.TP
.I MODULUS
String.
The prime modulus \fIp\fR in hex.
.TP
.I GENERATOR
String.
The generator \fIg\fR in hex.
.TP
.I CHECK
Boolean.
If \fBtrue\fR (the default) then check if \fIp\fR is prime.
.PP
A \fIdh closure\fR defines a group to be used for key exchange.

.SS logfile
\fBlogfile(\fIDICT\fB)\fR => \fIlog closure\fR
.PP
Valid keys in the \fIDICT\fR argument are:
.TP
.B filename
The path to log to.
.TP
.B class
A list of strings defining which classes of message to log.
The possible message classes are \fBdebug-config\fR,
\fBdebug-phase\fR, \fBdebug\fR, \fBinfo\fR, \fBnotice\fR,
\fBwarning\fR, \fBerror\fR, \fBsecurity\fR and \fBfatal\fR.
.IP
\fBall-debug\fR is the union of all the \fBdebug\fR... classes.
\fBdefault\fR is equivalent to \fBwarning, error, security, fatal\fR.
\fBverbose\fR is equivalent to \fBinfo, notice, warning, error,
security, fatal\fR.
\fBquiet\fR is equivalent to \fBfatal\fR.
.PP
A \fIlog closure\fR is a means of saving log messages.
See also \fBsyslog\fR below.

.SS makelist
\fBmakelist(\fIDICT\fB)\fR => \fILIST\fR
.PP
Returns the (flattened) list of values from the dictionary, discarding
the keys.

.SS map
\fBmap(\fICLOSURE\fB, \fIINPUT\fR...\fB)\fR => \fILIST\fR
.PP
Applies \fICLOSURE\fR to all its additional input arguments and
returns the resulting list.

.SS md5
\fBmd5\fR is a \fIhash closure\fR implementing the MD5 algorithm.

.SS null-netlink
\fBnull-netlink(\fIDICT\fB)\fR => \fInetlink closure\fR
.br
\fBnull-netlink(\fIDICT\fB)\fR => \fIpure closure\fR
.\" TODO pure closure is what it's called internally but this is a
.\" very opaque name to use in docs
.PP
Valid keys in the \fIDICT\fR argument are:
.TP
.B name
String.
The name for the netlink device.
The default is \fBnull-netlink\fR.
.TP
.B networks
List of strings.
The networks on the host side of the netlink device.
.TP
.B remote-networks
List of strings.
Networks that may be claimed by remote sites using this netlink device.
.TP
.B secnet-address
String.
IP address of this netlink.
Incompatible with \fBptp-address\fR.
.TP
.B ptp-address
String.
IP address of the other end of a point-to-point link.
Incompatible with \fBsecnet-address\fR.
.TP
.B mtu
Number.
The MTU of the netlink device.
The default is 1000.
.PP
If \fBptp-address\fR is used then the result is a \fInetlink closure\fR.
This can be used directly with the \fBlink\fR key in the \fBsites\fR
closure (see below).
.PP
If \fBsecnet-address\fR is used then the result is a \fIpure
closure\fR.
This must be evaluated to yield a \fInetlink closure\fR, using a
dictionary argument with the following keys:
.TP
.B routes
String list.
networks reachable via this tunnel, in \fIaddress\fB/\fIbits\fR format.
.TP
.B options
String list.
A list of options:
.RS
.TP
.B allow-route
Allow packets received via this tunnel to be routed down other tunnels
(without this option only packets from the host will be routed).
.TP
.B soft
Remove these routes from the host routing table when the link quality
is 0.
.RE
.TP
.B mtu
Number.
Default MTU over this link.
The default is inherited from the \fIpure closure\fR.
.TP
.B priority
Number.
The priority of this link.
Higher values beat lower values.
The default is 0.

.\" TODO ptp-address turns up in sites.conf, but why?  I think this
.\" is a bug in make-secnet-sites; it is not used by
 \" netlink_inst_create.

.PP
A \fInetlink closure\fR is a virtual IP link, and is supplied to the
\fBlink\fR key of a \fIsite\fR closure.
.PP
The netlink created by \fBnull-netlink\fR has no connection to the
host.
See \fBtun\fR and \fBuserv-ipif\fR below for more useful alternatives.



.SS randomfile
\fBrandomfile(\fIFILENAME\fR[\fB, \fIBLOCKING\fR]\fB)\fR => \fIrandomsource closure\fR
.TP
.I FILENAME
String.
Path to random device, e.g. \fI/dev/urandom\fR.
.TP
.I BLOCKING
Boolean.
\fBTrue\fR if this is a blocking device and \fBfalse\fR otherwise (the default).
Blocking device support is not implemented so this must always be
\fBFalse\fR or absent.
.PP
A \fIrandomsource closure\fR is a source of random numbers.

.SS readfile
\fBreadfile(\fIPATH\fB)\fR => \fISTRING\fR
.PP
Read the contents of the file \fIPATH\fR (a string) and return it as a string.

.SS eax-serpent

NEWS
* Planned for the future

Please note that the 0.1 series of secnet releases is now 'maintenance
only'; further development continues in secnet-0.2.

* Planned for version 0.1.17

Debconf support - if you are using the Debian packaged version and
your secnet configuration is autogenerated using debconf then the
upgrade to version 0.2.0 should just involve installing the package;
an appropriate 0.2-style configuration file will be generated
automatically.

* New in version 0.1.16

XXX XXX PROTOCOL COMPATIBILITY IS BROKEN BETWEEN VERSION 0.1.16 AND
XXX XXX ALL PREVIOUS VERSIONS.

Bugfix: rsa.c private-key now works properly when you choose not to
verify it.

Bugfix: serpent key setup was only using the first 8 bytes of the key
material. (Oops!)  Ian Jackson contributed a fix so the full 32 bytes
are used, in big-endian mode.

Debatable-bugfix: RSA operations now use PKCS1 v1.5-style padding

"Hacky parallelism" contributed by Ian Jackson; this permits
public-key operations to be performed in a subprocess during key
exchange, to make secnet more usable on very slow machines.  This is
not compiled in by default; if you find you need it (because key
exchanges are taking more than a second or two) then add
-DHACKY_PARALLEL to FLAGS in the Makefile.in and recompile.

udp module updates from Peter Benie:
  1) Handle the case where authbind-helper terminates with a signal
  2) Cope with signals being delivered during waitpid
  3) Add 'address' (optional) to the udp settings. This is an IP address
     that the socket will be bound to.
  4) Change the endianess of the arguments to authbind-helper.
     sprintf("%04X") already translates from machine repesentation to most
     significant octet first so htons reversed it again.

All uses of alloca() expunged by Peter Benie.

make-secnet-sites now supports configurations where each tunnel gets
its own interface on the host, and the IP router code in secnet is
disabled.  make-secnet-sites has been rewritten for clarity.  For
information on how to configure secnet for one-interface-per-tunnel,
see the example.conf file.

* New in version 0.1.15

Now terminates with an error when an "include" filename is not
specified in the configuration file (thanks to RJK).

RSA private key operations optimised using CRT. Thanks to SGT.

Now compiles cleanly with -Wwrite-strings turned on in gcc.

Anything sent to stderr once secnet has started running in the
background is now redirected to the system/log facility.

* New in version 0.1.14

The --help and --version options now send their output to stdout.

Bugfix: TUN flavour "BSD" no longer implies a BSD-style ifconfig and
route command invocation.  Instead "ioctl"-style is used, which should
work on both BSD and linux-2.2 systems.

If no "networks" parameter is specified for a netlink device then it
is assumed to be 0.0.0.0/0 rather than the empty set.  So, by default
there is a default route from each netlink device to the host machine.
The "networks" parameter can be used to implement a primitive
firewall, restricting the destination addresses of packets received
through tunnels; if a more complex firewall is required then implement
it on the host.

* New in version 0.1.13

site.c code cleaned up; no externally visible changes

secnet now calls setsid() after becoming a daemon.

secnet now supports TUN on Solaris 2.5 and above (and possibly other
STREAMS-based systems as well).

The TUN code now tries to auto-detect the type of "TUN" in use
(BSD-style, Linux-style or STREAMS-style).  If your configuration file
specifies "tun-old" then it defaults to BSD-style; however, since
"tun-old" will be removed in a future release, you should change your
configuration file to specify "tun" and if there's a problem also
specify the flavour in use.

Example:
netlink tun-old {
	...
};
should be rewritten as
netlink tun {
	flavour "bsd";
	...
};

The flavours currently defined are "bsd", "linux" and "streams".

The TUN code can now be configured to configure interfaces and
add/delete routes using one of several methods: invoking a
"linux"-style ifconfig/route command, a "bsd"-style ifconfig/route
command, "solaris-2.5"-style ifconfig/route command or calling ioctl()
directly.  These methods can be selected using the "ifconfig-type" and
"route-type" options.

Example:
netlink tun {
	ifconfig-type "ioctl";
	route-type "ioctl";
	...
};

The ioctl-based method is now the default for Linux systems.

Magic numbers used within secnet are now collected in the header file
"magic.h".

netlink now uses ICMP type=0 code=13 for 'administratively prohibited'
instead of code 9. See RFC1812 section 5.2.7.1.

The UDP comm module now supports a proxy server, "udpforward".  This
runs on a machine which is directly accessible by secnet and which can
send packets to appropriate destinations.  It's useful when the proxy
machine doesn't support source- and destination-NAT.  The proxy server
is specified using the "proxy" key in the UDP module configuration;
parameters are IP address (string) and port number.

Bugfix: ipset_to_subnet_list() in ipaddr.c now believed to work in all
cases, including 0.0.0.0/0

* New in version 0.1.12

IMPORTANT: fix calculation of 'now' in secnet.c; necessary for correct
operation.

(Only interesting for people building and modifying secnet by hand:
the Makefile now works out most dependencies automatically.)

The netlink code no longer produces an internal routing table sorted
by netmask length.  Instead, netlink instances have a 'priority'; the
table of routes is sorted by priority.  Devices like laptops that have
tunnels that must sometimes 'mask' parts of other tunnels should be
given higher priorities.  If a priority is not specified it is assumed
to be zero.

Example usage:
site laptop { ...
	link netlink {
		route "192.168.73.74/31";
		priority 10;
	};
};

* New in version 0.1.11

Lists of IP addresses in the configuration file can now include
exclusions as well as inclusions.  For example, you can specify all

secnet.c
/*
 * This file is part of secnet.
 * See README for full list of copyright holders.
 *
 * secnet is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 * 
 * secnet is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * version 3 along with secnet; if not, see
 * https://www.gnu.org/licenses/gpl.html.
 */

#include "secnet.h"
#include <stdio.h>
#include <assert.h>
#include <limits.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <unistd.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <pwd.h>
#include <grp.h>

#include "util.h"
#include "conffile.h"
#include "process.h"

#if __APPLE__
/* apple's poll() does not work on char devs */
# define USE_SELECT 1
#endif

/* XXX should be from autoconf */
static const char *configfile="/etc/secnet/secnet.conf";
static const char *sites_key="sites";
bool_t just_check_config=False;
static char *userid=NULL;
static uid_t uid=0;
static gid_t gid;
bool_t background=True;
static char *pidfile=NULL;
bool_t require_root_privileges=False;
cstring_t require_root_privileges_explanation=NULL;

static pid_t secnet_pid;

/* Structures dealing with poll() call */
struct poll_interest {
    beforepoll_fn *before; /* 0 if deregistered and waiting to be deleted */
    afterpoll_fn *after;
    void *state;
    int32_t nfds;
    cstring_t desc;
    LIST_ENTRY(poll_interest) entry;
};
static LIST_HEAD(, poll_interest) reg = LIST_HEAD_INITIALIZER(&reg);

static bool_t interest_isregistered(const struct poll_interest *i)
{
    return !!i->before;
}

static bool_t finished=False;

/* Parse the command line options */
static void parse_options(int argc, char **argv)
{
    int c;

    while (True) {
	int option_index = 0;
	static struct option long_options[] = {
	    {"verbose", 0, 0, 'v'},
	    {"nowarnings", 0, 0, 'w'},
	    {"help", 0, 0, 2},
	    {"version", 0, 0, 1},
	    {"nodetach", 0, 0, 'n'},
	    {"managed", 0, 0, 'm'},
	    {"silent", 0, 0, 'f'},
	    {"quiet", 0, 0, 'f'},
	    {"debug", 0, 0, 'd'},
	    {"config", 1, 0, 'c'},
	    {"just-check-config", 0, 0, 'j'},
	    {"sites-key", 1, 0, 's'},
	    {0,0,0,0}
	};

	c=getopt_long(argc, argv, "vwdnjc:ft:s:m",
		      long_options, &option_index);
	if (c==-1)
	    break;

	switch(c) {
	case 2:
	    /* Help */
	    printf("Usage: secnet [OPTION]...\n\n"
		   "  -f, --silent, --quiet   suppress error messages\n"
		   "  -w, --nowarnings        suppress warnings\n"
		   "  -v, --verbose           output extra diagnostics\n"
		   "  -c, --config=filename   specify a configuration file\n"
		   "  -j, --just-check-config stop after reading "
		   "configuration file\n"
		   "  -s, --sites-key=name    configuration key that "
		   "specifies active sites\n"
		   "  -n, --nodetach          do not run in background\n"
		   "  -m, --managed           running under a supervisor\n"
		   "  -d, --debug             output debug messages\n"
		   "      --help              display this help and exit\n"
		   "      --version           output version information "
		   "and exit\n"
		);
	    exit(0);
	    break;
      
	case 1:
	    /* Version */
	    printf("%s\n",version);
	    exit(0);
	    break;

	case 'd':
	    message_level|=M_DEBUG_CONFIG|M_DEBUG_PHASE|M_DEBUG;
	    /* fall through */
	case 'v':
	    message_level|=M_INFO|M_NOTICE|M_WARNING|M_ERR|M_SECURITY|
		M_FATAL;
	    break;

	case 'w':
	    message_level&=(~M_WARNING);
	    break;

	case 'f':
	    message_level=M_FATAL;
	    break;

	case 'n':
	    background=False;
	    break;

	case 'm':
	    secnet_is_daemon=True;
	    break;

	case 'c':
	    if (optarg)
		configfile=safe_strdup(optarg,"config_filename");
	    else
		fatal("secnet: no config filename specified");
	    break;

	case 'j':
	    just_check_config=True;
	    break;

	case 's':
	    if (optarg)
		sites_key=safe_strdup(optarg,"sites-key");
	    else
		fatal("secnet: no sites key specified");
	    break;

	case '?':
	    exit(1);
	    break;

	default:
	    Message(M_ERR,"secnet: Unknown getopt code %c\n",c);
	}
    }

    if (argc-optind != 0) {
	Message(M_ERR,"secnet: You gave extra command line parameters, "
		"which were ignored.\n");
    }
}

static void setup(dict_t *config)
{
    list_t *l;
    item_t *site;
    dict_t *system;
    struct passwd *pw;
    struct cloc loc;
    int i;

    l=dict_lookup(config,"system");

    if (!l || list_elem(l,0)->type!=t_dict) {
	fatal("configuration does not include a \"system\" dictionary");
    }
    system=list_elem(l,0)->data.dict;
    loc=list_elem(l,0)->loc;

    /* Arrange systemwide log facility */
    l=dict_lookup(system,"log");
    if (!l) {
	fatal("configuration does not include a system/log facility");
    }
    system_log=init_log(l);

    /* Who are we supposed to run as? */
    userid=dict_read_string(system,"userid",False,"system",loc);
    if (userid) {
	if (!(pw=getpwnam(userid)))
	    fatal("userid \"%s\" not found",userid);
	uid=pw->pw_uid;
	gid=pw->pw_gid;
    }

    /* Pidfile name */
    pidfile=dict_read_string(system,"pidfile",False,"system",loc);

    /* Check whether we need root privileges */
    if (require_root_privileges && uid!=0) {
	fatal("the configured feature \"%s\" requires "
	      "that secnet retain root privileges while running.",
	      require_root_privileges_explanation);
    }

    /* Go along site list, starting sites */
    l=dict_lookup(config,sites_key);
    if (!l) {
	Message(M_WARNING,"secnet: configuration key \"%s\" is missing; no "
		"remote sites are defined\n",sites_key);
    } else {
	i=0;
	while ((site=list_elem(l, i++))) {
	    struct site_if *s;
	    if (site->type!=t_closure) {
		cfgfatal(site->loc,"system","non-closure in site list");
	    }
	    if (site->data.closure->type!=CL_SITE) {
		cfgfatal(site->loc,"system","non-site closure in site list");
	    }
	    s=site->data.closure->interface;
	    s->control(s->st,True);
	}
    }
}

struct poll_interest *register_for_poll(void *st, beforepoll_fn *before,
		       afterpoll_fn *after, cstring_t desc)
{
    struct poll_interest *i;

    NEW(i);
    i->before=before;
    i->after=after;
    i->state=st;
    i->nfds=0;
    i->desc=desc;
    LIST_INSERT_HEAD(&reg, i, entry);
    return i;
}

void deregister_for_poll(struct poll_interest *i)
{
    /* We cannot simply throw this away because we're reentrantly
     * inside the main loop, which needs to remember which range of
     * fds corresponds to this now-obsolete interest */
    i->before=0;
}

static void system_phase_hook(void *sst, uint32_t newphase)
{
    if (newphase==PHASE_SHUTDOWN && pidfile) {
	/* Try to unlink the pidfile; don't care if it fails */
	unlink(pidfile);
    }
}

#if USE_SELECT
static int fakepoll(struct pollfd *fds, int nfds, int timeout) {
    fd_set infds[1], outfds[1];
    int maxfd = -1, i, rc;
    struct timeval tvtimeout;
    FD_ZERO(infds);
    FD_ZERO(outfds);
    for(i = 0; i < nfds; ++i) {
	if(fds[i].events & POLLIN)
	    FD_SET(fds[i].fd, infds);
	if(fds[i].events & POLLOUT)
	    FD_SET(fds[i].fd, outfds);
	if(fds[i].fd > maxfd)
	    maxfd = fds[i].fd;
    }
    if(timeout != -1) {
	tvtimeout.tv_sec = timeout / 1000;
	tvtimeout.tv_usec = 1000 * (timeout % 1000);
    }
    rc = select(maxfd + 1, infds, outfds, NULL, 
		timeout == -1 ? NULL : &tvtimeout);
    if(rc >= 0) {
	for(i = 0; i < nfds; ++i) {
	    int revents = 0;
	    if(FD_ISSET(fds[i].fd, infds))
		revents |= POLLIN;
	    if(FD_ISSET(fds[i].fd, outfds))
		revents |= POLLOUT;
	    fds[i].revents = revents;
	}
    }
    return rc;
}
#endif

struct timeval tv_now_global;
uint64_t now_global;

static void run(void)
{
    struct poll_interest *i, *itmp;
    int rv, nfds, idx;
    int timeout;
    struct pollfd *fds=0;
    int allocdfds=0, shortfall=0;

    Message(M_NOTICE,"%s [%d]: starting\n",version,secnet_pid);

    do {
	if (gettimeofday(&tv_now_global, NULL)!=0) {
	    fatal_perror("main loop: gettimeofday");
	}
	now_global=((uint64_t)tv_now_global.tv_sec*(uint64_t)1000)+
	           ((uint64_t)tv_now_global.tv_usec/(uint64_t)1000);
	idx=0;
	LIST_FOREACH(i, &reg, entry) {
	    int check;
	    if (interest_isregistered(i)) {
		for (check=0; check<i->nfds; check++) {
		    if(fds[idx+check].revents & POLLNVAL) {
			fatal("run: poll (%s#%d) set POLLNVAL", i->desc, check);
		    }
		}
		i->after(i->state, fds+idx, i->nfds);
	    }
	    idx+=i->nfds;
	}
	if (shortfall) {
	    allocdfds *= 2;
	    allocdfds += shortfall;
	    REALLOC_ARY(fds,allocdfds);
	}
	shortfall=0;
	idx=0;
	timeout=-1;
	LIST_FOREACH_SAFE(i, &reg, entry, itmp) {
	    int remain=allocdfds-idx;
	    nfds=remain;
	    if (interest_isregistered(i)) {
		rv=i->before(i->state, fds+idx, &nfds, &timeout);
		if (rv!=0) {
		    if (rv!=ERANGE)
			fatal("run: beforepoll_fn (%s) returns %d",i->desc,rv);
		    assert(nfds < INT_MAX/4 - shortfall);
		    shortfall += nfds-remain;
		    nfds=0;
		    timeout=0;
		}
	    } else {
		nfds=0;
	    }
	    if (timeout<-1) {
		fatal("run: beforepoll_fn (%s) set timeout to %d",
		      i->desc,timeout);
	    }
	    if (!interest_isregistered(i)) {
		/* check this here, rather than earlier, so that we
		   handle the case where i->before() calls deregister */
		LIST_REMOVE(i, entry);
		free(i);
		continue;
	    }
	    idx+=nfds;
	    i->nfds=nfds;
	}
	do {
	    if (finished) break;
#if USE_SELECT
	    rv=fakepoll(fds, idx, timeout);
#else
	    rv=poll(fds, idx, timeout);
#endif
	    if (rv<0) {
		if (errno!=EINTR) {
		    fatal_perror("run: poll");
		}
	    }
	} while (rv<0);
    } while (!finished);
    free(fds);
}

bool_t will_droppriv(void)
{
    assert(current_phase >= PHASE_SETUP);
    return !!uid;
}

/* Surrender privileges, if necessary */
static void droppriv(void)
{
    if (userid) {
	if (setgid(gid)!=0)
	    fatal_perror("can't set gid to %ld",(long)gid);
	if (initgroups(userid, gid) < 0)
	    fatal_perror("initgroups");	
	if (setuid(uid)!=0) {
	    fatal_perror("can't set uid to \"%s\"",userid);
	}
	assert(getuid() == uid);
	assert(geteuid() == uid);
	assert(getgid() == gid);
	assert(getegid() == gid);
    }
}

/* Become a daemon, if necessary */
static void become_daemon(void)
{
    FILE *pf=NULL;
    pid_t p;
    int errfds[2];

    add_hook(PHASE_SHUTDOWN,system_phase_hook,NULL);

    /* We only want to become a daemon if we are not one
     already */
    if (background && !secnet_is_daemon) {
	p=fork();
	if (p>0) {
	    /* Parent process - just exit */
	    _exit(0);
	} else if (p==0) {
	    /* Child process - all done, just carry on */
	    secnet_is_daemon=True;
	    if (setsid() < 0)
		fatal_perror("setsid");
	} else {
	    /* Error */
	    fatal_perror("cannot fork");
	    exit(1);
	}
    }
    if (secnet_is_daemon) {
	/* stderr etc are redirected to the system/log facility */
	pipe_cloexec(errfds);
	if (dup2(errfds[1],0) < 0
	    || dup2(errfds[1],1) < 0
	    || dup2(errfds[1],2) < 0)
	    fatal_perror("can't dup2 pipe");
	if (close(errfds[1]) < 0)
	    fatal_perror("can't close redundant pipe endpoint");
	log_from_fd(errfds[0],"stderr",system_log);
    }
    secnet_pid=getpid();
    
    /* Now we can write the pidfile */
    if (pidfile) {
	pf=fopen(pidfile,"w");
	if (!pf) {
	    fatal_perror("cannot open pidfile \"%s\"",pidfile);
	}
	if (fprintf(pf,"%ld\n",(long)secnet_pid) < 0
	    || fclose(pf) < 0)
	    fatal_perror("cannot write to pidfile \"%s\"",pidfile);
    }
}

static signal_notify_fn finish,ignore_hup;
static void finish(void *st, int signum)
{
    finished=True;
    Message(M_NOTICE,"%s [%d]: received %s\n",version,secnet_pid,(string_t)st);
}
static void ignore_hup(void *st, int signum)
{
    Message(M_INFO,"%s [%d]: received SIGHUP\n",version,secnet_pid);
    return;
}

int main(int argc, char **argv)
{
    dict_t *config;

    phase_hooks_init();

    enter_phase(PHASE_GETOPTS);
    parse_options(argc,argv);

    enter_phase(PHASE_READCONFIG);
    config=read_conffile(configfile);

    enter_phase(PHASE_SETUP);
    setup(config);

    if (just_check_config) {
	Message(M_INFO,"configuration file check complete\n");
	exit(0);
    }

    enter_phase(PHASE_DAEMONIZE);
    become_daemon();
    
    enter_phase(PHASE_GETRESOURCES);
    /* Appropriate phase hooks will have been run */
    
    enter_phase(PHASE_DROPPRIV);
    droppriv();

    start_signal_handling();
    request_signal_notification(SIGTERM,finish,safe_strdup("SIGTERM","run"));
    if (!background) request_signal_notification(SIGINT,finish,
						 safe_strdup("SIGINT","run"));
    request_signal_notification(SIGHUP,ignore_hup,NULL);
    enter_phase(PHASE_RUN);
    run();

    enter_phase(PHASE_SHUTDOWN);
    Message(M_NOTICE,"%s [%d]: finished\n",version,secnet_pid);

    return 0;
}
